File: utilities/GoDBTool/src_tool/dialect/mysql/MySql.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Data;
using System.Data.SqlClient;
using System.Drawing;
using Npgsql;
using Npgsql.Logging;
using System.IO;
using src_tool.templates;
using MySql.Data.MySqlClient;


/**
pt mssql: se dezactiveaza indecsii cand stergem chestii mari
  ALTER INDEX ALL ON schema_name.table_name DISABLE;

ALTER INDEX ALL ON schema_name.table_name REBUILD WITH (ONLINE = ON);

pt mysql

ALTER TABLE table_name DISABLE KEYS;
ALTER TABLE table_name ENABLE KEYS;
SET FOREIGN_KEY_CHECKS=0;
SET FOREIGN_KEY_CHECKS=1;

pentru a investiga paramentrii procedurilor stocate
SELECT * from INFORMATION_SCHEMA.PARAMETERS




*/
namespace src_tool
{
    public partial class MySqlDialect : GenericDialect
    {
        public MySqlConnection connection;

        public override string getSql()
        { 
            return "mysql";
        }
    
        public override string SqlSeparator()
        { 
            return ";";
        }
        public string tokenizIdentif(string identif)
        {
            return $"`{identif}`";
        }
        public string tokenizTable(DbTable table)
        {
            return tokenizIdentif(table.SqlTableNameModel);
        }
        public override string dropTable(DbTable table)
        {
            var s = $@"DROP TABLE IF EXISTS {tokenizIdentif(table.SqlTableNameModel)}";
            return s;
        }
        public override string addTable(DbTable table)
        {
            var NL = 
                @"
                ";
            var constraintorList = new List<string>();
            var sqlTableName = table.SqlTableNameModel;

            var columnsArr = new List<string>();
            foreach( var col in table.columns)
            {
                columnsArr.Add( this._addColumn(col, ref constraintorList) );
            }
            var columnsDefs = string.Join( ","+NL, columnsArr );
            var colID_Name = table.PrimaryColumn.sqlName;

            var constraintors = string.Join( ","+NL, constraintorList );
            if( constraintors != "")
                constraintors = $",{NL}{constraintors}";

            var tableAdd = $@"        
            #-------------------------------------------------------------------	
            CREATE TABLE IF NOT EXISTS {tokenizIdentif(sqlTableName)}
            (
                { columnsDefs}

                ,PRIMARY KEY ({tokenizIdentif(colID_Name)})
                {constraintors}
            ) ENGINE=InnoDB";

                return tableAdd;
        }
        public override string dropColumn(DbTable table,DbColumn column)
        {
            var s = $@"
            ALTER TABLE {tokenizIdentif(table.SqlTableNameModel)}
            DROP COLUMN {tokenizIdentif(column.sqlName)}";
            return s;
        }
        public override string addColumn(DbTable table, DbColumn column)
        { 
            var listConstraintors = new  List<string>();
            var colData = _addColumn( column, ref listConstraintors);
            var s = $@"
            ALTER TABLE {tokenizIdentif(table.SqlTableNameModel)}
            ADD COLUMN {colData}";
            return s;
        }
        public string _addColumn(DbColumn column, ref List<string> listConstraintors)
        {
            var colName = column.sqlName;
            if( column.bIsIdentity )
            {
                if( column.langType == "int64")
                {
                    return $"{tokenizIdentif(colName)} BIGINT NOT NULL AUTO_INCREMENT";
                }else
                {
                    return $"{tokenizIdentif(colName)} INT NOT NULL AUTO_INCREMENT";
                }
            }else
            {
                var sqlType = column.sqlType;
                if( column.ForeignKey != null)
                {
                    var targetTableSqlName = tokenizTable(column.ForeignKey);
                    var targetTable_ID     = tokenizIdentif(column.ForeignKey.PrimaryColumn.sqlName);

                    var fkName      = tokenizIdentif(column.sqlName);
                    var fk_fkName   = tokenizIdentif($"fk_{column.sqlName}");

                    listConstraintors.Add($"CONSTRAINT {fk_fkName} FOREIGN KEY ({fkName}) REFERENCES {targetTableSqlName}({targetTable_ID})");
                    //listConstraintors.Add($"CONSTRAINT {fk_fkName} FOREIGN KEY ({fkName}) REFERENCES {targetTableSqlName}({targetTable_ID}) ON DELETE RESTRICT");
                    
                    return $"{fkName} INT NULL";
                }else
                {
                    if( column.bIsNullable )
                    {
                        return $"{tokenizIdentif(colName)} {sqlType} NULL";
                    }else
                    {
                        return $"{tokenizIdentif(colName)} {sqlType} NOT NULL";
                        //return $"{tokenizIdentif(colName)} {sqlType} NOT NULL DEFAULT ''::character varying";                        
                    }
                }
            }
        }
        public override string updateTable(DbTable table, DbTable tablePrev)
        {
            if( table.SqlTableNameModel != tablePrev.SqlTableNameModel)
            {
                var s = $@"
                RENAME TABLE {tokenizIdentif(tablePrev.SqlTableNameModel)}
                TO {tokenizIdentif(table.SqlTableNameModel)}";
                return s;
            }
            return "";
        }
        //----------------------------------------------------
        public override string updateColumn(DbTable table, DbColumn column, DbColumn columnPrev)
        {
            if( columnPrev.langName != column.langName)
                return printError("internal error");

            if( column.sqlType != columnPrev.sqlType )
            {
                var r = $@"
                ALTER TABLE {tokenizIdentif(table.SqlTableNameModel)} 
                MODIFY {tokenizIdentif(column.sqlName)} {column.sqlType}";
                return r;
            }

            if( column.sqlName != columnPrev.sqlName)
            {
                var s = $@"
                ALTER TABLE {tokenizIdentif(table.SqlTableNameModel)}
                CHANGE COLUMN {tokenizIdentif(columnPrev.sqlName)} {tokenizIdentif(column.sqlName)} {column.sqlType}";
                return s;
            }

            return "";
        }
        public override string addFKConstrictor(DbTable table,DbColumn column)
        {
            var tablename   = tokenizTable(column.ForeignKey);
            var tableID     = tokenizIdentif(column.ForeignKey.PrimaryColumn.sqlName);

            var myTableName = tokenizTable(table);
            var fkName      = tokenizIdentif(column.sqlName);
            var fk_fkName   = tokenizIdentif($"fk_{column.sqlName}");
            
            var s = $@"
                ALTER TABLE {myTableName}
                ADD CONSTRAINT {fk_fkName} FOREIGN KEY ({fkName}) REFERENCES {tablename} ({tableID});";
            return s;
        }
        public override string dropFKConstrictor(DbTable table,DbColumn column)
        {
            //var tablename   = tokenizIdentif(column.ForeignKey.SqlTableNameModel);
            //var tableID     = tokenizIdentif(column.ForeignKey.PrimaryColumn.sqlName);

            var myTableName = tokenizTable(table);
            //var fkName      = tokenizIdentif(column.sqlName);
            var fk_fkName   = tokenizIdentif($"fk_{column.sqlName}");

            var s = $@"
                ALTER TABLE {myTableName}
                DROP FOREIGN KEY {fk_fkName};";
            return s;
        }
        public override string getSqlType( string langType, ref bool bIsNullable, string nameOfColumn)
        {
            langType = cleanNameGoStruct(langType);
            if(langType.StartsWith("[]"))
            {
                return printError($"type {langType} not supported!");
            }

            switch( langType )
            { 
                case "NullString":      bIsNullable= true; return isLongType(nameOfColumn)?"TEXT":"VARCHAR(255)";
                case "NullBool":        bIsNullable= true; return "BOOLEAN";

                case "NullByte":        bIsNullable= true; return "SMALLINT";
                case "NullInt16":       bIsNullable= true; return "SMALLINT";
                case "NullInt":         bIsNullable= true; return "INT";
                case "NullInt32":       bIsNullable= true; return "INT";
                case "NullInt64":       bIsNullable= true; return "BIGINT";
                case "NullFloat64":         bIsNullable= true; return "double precision";
                case "time.NullTime":   bIsNullable= true; return "TIMESTAMP";
                case "NullTime":        bIsNullable= true; return "TIMESTAMP";

                case "string":      return isLongType(nameOfColumn)?"TEXT":"VARCHAR(255)";
                case "char":        return "CHAR";
                case "bool":        return "BOOLEAN";

                case "byte":        return "SMALLINT";
                case "int16":       return "SMALLINT";
                case "int":         return "INT";
                case "int32":       return "INT";
                case "int64":       return "BIGINT";

                //case "time.Time":   return "TIMESTAMP";
                case "Time":        return "TIMESTAMP";

                case "UUID": return "UUID";
                case "uuid.UUID": return "UUID";

                case "float32": return "FLOAT";
                case "float64": return "DOUBLE PRECISION";
            }
            return printError($"not supported type {langType}");
        }
        public override string getProperty(  string propName, 
            string tableName, string colName, string colValue)
        {
            try
            {
                var requestText = $@"
                SELECT {tokenizIdentif(colValue)} FROM {tokenizIdentif(tableName)}
                 WHERE {tokenizIdentif(colName)} = '{propName}'
            ";

            using( var cmd = new MySqlCommand( requestText, this.connection))
            {
                var obj = cmd.ExecuteScalar();
                if( obj != null)
                    return obj.ToString();
            }
            }catch(Exception e )
            {
                var s = e.Message;
            }
            return null;
        }
        public override void updateProperty( 
            string propName, string value, 
            string tableName, string colName, string colValue)
        {
            var requestText = $@"

            UPDATE {tokenizIdentif(tableName)}
            SET   {tokenizIdentif(colValue)}  = '{value}'
            WHERE {tokenizIdentif(colName)}   = '{propName}'
            ";

            using( var cmd = new MySqlCommand( requestText, this.connection))
            {
                var reader = cmd.ExecuteNonQuery();
            }
        }
        public override void insertProperty(
            string propName, string value, 
            string tableName, string colName, string colValue)
        {
            var requestText = $@"
            INSERT INTO {tokenizIdentif(tableName)} ( {tokenizIdentif("ID")}, {tokenizIdentif(colName)}, {tokenizIdentif(colValue)}) 
            VALUES (  default, '{propName}', '{value}' ) ";

            using( var cmd = new MySqlCommand ( requestText, this.connection))
            {
                var reader = cmd.ExecuteNonQuery();
            }
        }
        public override void execScript( string scriptTxt)
        {
            var parts = scriptTxt.Split(';');
            var iPart = 0;
            foreach( var part in parts)
            {
                var script1 = part.Trim();
                if( script1.Length >  0 )
                {
                    using( var cmd = new MySqlCommand( script1, this.connection))
                    {
                        try
                        {
                            var reader = cmd.ExecuteNonQuery();
                        }catch(Exception e)
                        {
                            var msg = e.InnerException != null ? e.InnerException.Message : e.Message;
                            Console.WriteLine($"Error exec script part {iPart}: {msg}");
                            throw e;
                        }
                    }
                }
                iPart ++;
            }
        }

        public override bool startConnection(GenericDialectArg arg)
        {
            MySqlConnection cnn = null;
            try
            {
                var connectionString = arg.connection_string;
                cnn = new MySqlConnection(connectionString);
                cnn.Open();
                this.connection = cnn;

                if( true)
                {
                    this.fnProcessData( this, arg);
                    return true;
                }
                else
                {
                    using(var transaction = cnn.BeginTransaction())
                    {
                        try
                        {
                            this.fnProcessData( this, arg);
                            transaction.Commit();
                            return true;
                        }catch( Exception e )
                        {
                            transaction.Rollback();

                            Console.WriteLine(e.InnerException != null? e.InnerException.Message: e.Message);
                            Console.WriteLine("Transaction is Canceled!!");

                            if( cnn != null)
                                cnn.Close();
                            return false;
                        }
                    }                
                }
                
            }catch(Exception e )
            {
                Console.WriteLine(e.Message);
            }finally
            {
                if( cnn != null)
                    cnn.Close();
            }
            return false;
        }
    }
}

File: library/tools/gomigration/src/dialect/mysql/MySql.go
package gomigration_mysql

import (
	"database/sql"
	"fmt"
	"strings"

	dialect "github.com/bbitere/atomicsql_golang.git/tools/gomigration/src/dialect"
	utils "github.com/bbitere/atomicsql_golang.git/tools/gomigration/src/utils"

	_ "github.com/go-sql-driver/mysql"
)

type MySqlDialect struct {
	connection *sql.DB
    Base1 dialect.TGenericDialect;
}

// GetSql returns the SQL dialect used by MySQL
func (m *MySqlDialect) GetSql() string {
	return "mysql"
}

// SqlSeparator returns the SQL separator used by MySQL
func (m *MySqlDialect) SqlSeparator() string {
	return ";"
}

// TokenizeIdentif tokenizes an identifier for MySQL
func (m *MySqlDialect) TokenizeIdentif(identif string) string {
	return fmt.Sprintf("`%s`", identif)
}

// TokenizeTable tokenizes a table name for MySQL
func (m *MySqlDialect) TokenizeTable(table *dialect.DbTable) string {
	return m.TokenizeIdentif(table.SqlTableNameModel)
}

// DropTable generates a DROP TABLE statement for MySQL
func (m *MySqlDialect) DropTable(table *dialect.DbTable) string {
	return fmt.Sprintf("DROP TABLE IF EXISTS %s", m.TokenizeIdentif(table.SqlTableNameModel))
}

// AddTable generates a CREATE TABLE statement for MySQL
func (m *MySqlDialect) AddTable(table *dialect.DbTable) string {
	var NL = "\n"
	var constraintorList []string
	var columnsArr []string
	sqlTableName := table.SqlTableNameModel

	for _, col := range table.Columns {
		columnsArr = append(columnsArr, m._AddColumn(col, &constraintorList))
	}

	columnsDefs := strings.Join(columnsArr, ","+NL)
	colIDName := table.PrimaryColumn.SqlName

	constraintors := strings.Join(constraintorList, ","+NL)
	if constraintors != "" {
		constraintors = fmt.Sprintf(",%s%s", NL, constraintors)
	}

	tableAdd := fmt.Sprintf(`
            #-------------------------------------------------------------------	
            CREATE TABLE IF NOT EXISTS %s
            (
                %s

                ,PRIMARY KEY (%s)
                %s
            ) ENGINE=InnoDB`, m.TokenizeIdentif(sqlTableName), columnsDefs, m.TokenizeIdentif(colIDName), constraintors)

	return tableAdd
}

// DropColumn generates a DROP COLUMN statement for MySQL
func (m *MySqlDialect) DropColumn(table *dialect.DbTable, column *dialect.DbColumn) string {
	return fmt.Sprintf(`
            ALTER TABLE %s
            DROP COLUMN %s`, m.TokenizeIdentif(table.SqlTableNameModel), m.TokenizeIdentif(column.SqlName))
}

// AddColumn generates an ADD COLUMN statement for MySQL
func (m *MySqlDialect) AddColumn(table *dialect.DbTable, column *dialect.DbColumn) string {
	listConstraintors := make([]string, 0)
	colData := m._AddColumn(column, &listConstraintors)
	return fmt.Sprintf(`
            ALTER TABLE %s
            ADD COLUMN %s`, m.TokenizeIdentif(table.SqlTableNameModel), colData)
}

// _AddColumn generates the column definition for MySQL
func (m *MySqlDialect) _AddColumn(column *dialect.DbColumn, listConstraintors *[]string) string {
	colName := column.SqlName

	if column.IsIdentity {
		if column.LangType == "int64" {
			return fmt.Sprintf("%s BIGINT NOT NULL AUTO_INCREMENT", m.TokenizeIdentif(colName))
		} else {
			return fmt.Sprintf("%s INT NOT NULL AUTO_INCREMENT", m.TokenizeIdentif(colName))
		}
	} else {
		sqlType := column.SqlType
		if column.ForeignKey != nil {
			targetTableSqlName := m.TokenizeTable(column.ForeignKey)
			targetTableID := m.TokenizeIdentif(column.ForeignKey.PrimaryColumn.SqlName)

			fkName := m.TokenizeIdentif(column.SqlName)
			fkFKName := m.TokenizeIdentif(fmt.Sprintf("fk_%s", column.SqlName))

			*listConstraintors = append(*listConstraintors, fmt.Sprintf("CONSTRAINT %s FOREIGN KEY (%s) REFERENCES %s(%s)", fkFKName, fkName, targetTableSqlName, targetTableID))

			return fmt.Sprintf("%s INT NULL", fkName)
		} else {
			if column.IsNullable {
				return fmt.Sprintf("%s %s NULL", m.TokenizeIdentif(colName), sqlType)
			} else {
				return fmt.Sprintf("%s %s NOT NULL", m.TokenizeIdentif(colName), sqlType)
			}
		}
	}
}

// UpdateTable generates an UPDATE TABLE statement for MySQL
func (m *MySqlDialect) UpdateTable(table *dialect.DbTable, tablePrev *dialect.DbTable) string {
	if table.SqlTableNameModel != tablePrev.SqlTableNameModel {
		return fmt.Sprintf(`
                RENAME TABLE %s
                TO %s`, m.TokenizeIdentif(tablePrev.SqlTableNameModel), m.TokenizeIdentif(table.SqlTableNameModel))
	}
	return ""
}

// UpdateColumn generates an UPDATE COLUMN statement for MySQL
func (m *MySqlDialect) UpdateColumn(table *dialect.DbTable, column *dialect.DbColumn, columnPrev *dialect.DbColumn) string {
	if columnPrev.LangName != column.LangName {
		return m.Base1.PrintError("internal error")
	}

	if column.SqlType != columnPrev.SqlType {
		return fmt.Sprintf(`
                ALTER TABLE %s 
                MODIFY %s %s`, m.TokenizeIdentif(table.SqlTableNameModel), m.TokenizeIdentif(column.SqlName), column.SqlType)
	}

	if column.SqlName != columnPrev.SqlName {
		return fmt.Sprintf(`
                ALTER TABLE %s
                CHANGE COLUMN %s %s %s`, m.TokenizeIdentif(table.SqlTableNameModel), m.TokenizeIdentif(columnPrev.SqlName), m.TokenizeIdentif(column.SqlName), column.SqlType)
	}

	return ""
}

// AddFKConstraint generates an ADD FOREIGN KEY statement for MySQL
func (m *MySqlDialect) AddFKConstrictor(table *dialect.DbTable, column *dialect.DbColumn) string {
	tableName := m.TokenizeTable(column.ForeignKey)
	tableID := m.TokenizeIdentif(column.ForeignKey.PrimaryColumn.SqlName)

	myTableName := m.TokenizeTable(table)
	fkName := m.TokenizeIdentif(column.SqlName)
	fkFKName := m.TokenizeIdentif(fmt.Sprintf("fk_%s", column.SqlName))

	return fmt.Sprintf(`
                ALTER TABLE %s
                ADD CONSTRAINT %s FOREIGN KEY (%s) REFERENCES %s (%s);`, myTableName, fkFKName, fkName, tableName, tableID)
}

// DropFKConstraint generates a DROP FOREIGN KEY statement for MySQL
func (m *MySqlDialect) DropFKConstrictor(table *dialect.DbTable, column *dialect.DbColumn) string {
	myTableName := m.TokenizeTable(table)
	fkFKName := m.TokenizeIdentif(fmt.Sprintf("fk_%s", column.SqlName))

	return fmt.Sprintf(`
                ALTER TABLE %s
                DROP FOREIGN KEY %s;`, myTableName, fkFKName)
}

// GetSqlType returns the equivalent MySQL data type for a given Go type
func (m *MySqlDialect) GetSqlType(langType string, bIsNullable *bool, nameOfColumn string) string {
	langType = m.Base1.CleanNameGoStruct(langType)
	if strings.HasPrefix(langType, "[]") {
		return m.Base1.PrintError(fmt.Sprintf("type %s not supported!", langType))
	}

	switch langType {
	case "NullString":
		*bIsNullable = true
		return m.IsLongType(nameOfColumn, "TEXT", "VARCHAR(255)" )
	case "NullBool":
		*bIsNullable = true
		return "BOOLEAN"
	case "NullByte", "NullInt16", "NullInt", "NullInt32":
		*bIsNullable = true
		return "SMALLINT"
	case "NullInt64":
		*bIsNullable = true
		return "BIGINT"
	case "NullFloat64":
		*bIsNullable = true
		return "DOUBLE PRECISION"
	case "time.NullTime", "NullTime":
		*bIsNullable = true
		return "TIMESTAMP"
	case "string":
		return m.IsLongType(nameOfColumn, "TEXT", "VARCHAR(255)" )
	case "char":
		return "CHAR"
	case "bool":
		return "BOOLEAN"
	case "byte", "int16":
		return "SMALLINT"
	case "int", "int32":
		return "INT"
	case "int64":
		return "BIGINT"
	case "Time":
		return "TIMESTAMP"
	case "UUID", "uuid.UUID":
		return "UUID"
	case "float32":
		return "FLOAT"
	case "float64":
		return "DOUBLE PRECISION"
	default:
		return m.Base1.PrintError(fmt.Sprintf("not supported type %s", langType))
	}
}


// getProperty retrieves a property from the database for a specific table, column, and value
func (m *MySqlDialect) GetProperty(propName, tableName, colName, colValue string) string {
	requestText := fmt.Sprintf(`
                SELECT %s FROM %s
                WHERE %s = '%s'`, m.TokenizeIdentif(colValue), m.TokenizeIdentif(tableName), m.TokenizeIdentif(colName), propName)

	row := m.connection.QueryRow(requestText)
	var value string
	if err := row.Scan(&value); err != nil {
		return ""
	}
	return value
}

// UpdateProperty updates a property in the database for a specific table, column, and value
func (m *MySqlDialect) UpdateProperty(propName, value, tableName, colName, colValue string) {
	requestText := fmt.Sprintf(`
                UPDATE %s
                SET   %s  = '%s'
                WHERE %s   = '%s'`, m.TokenizeIdentif(tableName), m.TokenizeIdentif(colValue), value, m.TokenizeIdentif(colName), propName)

	_, err := m.connection.Exec(requestText)
	if err != nil {
		fmt.Println("Error updating property:", err)
	}
}

// InsertProperty inserts a property into the database for a specific table, column, and value
func (m *MySqlDialect) InsertProperty(propName, value, tableName, colName, colValue string) {
	requestText := fmt.Sprintf(`
                INSERT INTO %s ( %s, %s, %s) 
                VALUES (  default, '%s', '%s' )`, m.TokenizeIdentif(tableName), m.TokenizeIdentif("ID"), m.TokenizeIdentif(colName), m.TokenizeIdentif(colValue), propName, value)

	_, err := m.connection.Exec(requestText)
	if err != nil {
		fmt.Println("Error inserting property:", err)
	}
}

// ExecScript executes a SQL script in MySQL
func (m *MySqlDialect) ExecScript(scriptTxt string) {
	parts := strings.Split(scriptTxt, ";")
	for i, part := range parts {
		script1 := strings.TrimSpace(part)
		if len(script1) > 0 {
			_, err := m.connection.Exec(script1)
			if err != nil {
				msg := err.Error()
				fmt.Printf("Error exec script part %d: %s\n", i, msg)
				panic(err)
			}
		}
	}
}

// StartConnection initiates a connection to the MySQL database
func (m *MySqlDialect) StartConnection(arg dialect.IGenericDialectArg) bool {
	var err error

    var connString = arg.GetGenericDialectArg().Connection_String;

    var connStr = utils.Utils_parseConnectionString(connString);
	var	dataSource = fmt.Sprintf(
			"Data Source=%s;"+
				"Initial Catalog=%s;"+
				"User id=%s;"+
				"password=%s;",
			connStr.Host,
			connStr.DbName,
			connStr.User,
			connStr.Password)

	m.connection, err = sql.Open("mysql", dataSource )
	if err != nil {
		fmt.Println("Error opening database connection:", err)
		return false
	}
	defer m.connection.Close()

	err = m.connection.Ping()
	if err != nil {
		fmt.Println("Error pinging database:", err)
		return false
	}

	if true {
		m.Base1.FnProcessData( m, arg)
		return true
	}

	return false
}



func (pd *MySqlDialect) IsLongType(sqlColumnName string, type1 string, type2 string) string {

    return pd.Base1.IsLongType(sqlColumnName, type1, type2);
}

func (pd *MySqlDialect) PrintError(err string) string {

    return pd.Base1.PrintError(err)
}
func (pd *MySqlDialect) SetCurrentFile(file string)  {
    
    pd.Base1.SetCurrentFile(file)
}

func (pd *MySqlDialect) CleanNameGoStruct(name string) string {

    return pd.Base1.CleanNameGoStruct(name )
}
func (pd *MySqlDialect) GetGenericDialect() *dialect.TGenericDialect{

    return &pd.Base1;
}